scvi/external/tangram/_model.py

```diff
@@ -0,0 +1,356 @@
+        )
+        self._model_summary_string = (
+            "TangramMapper Model with params: \nn_obs_sc: {}, n_obs_sp: {}"
+        ).format(
+            self.n_obs_sc,
+            self.n_obs_sp,
+        )
+        self.init_params_ = self._get_init_params(locals())
+
+    def get_mapper_matrix(self) -> np.ndarray:
+        """
+        Return the mapping matrix.
+
+        Returns
+        -------
+        Mapping matrix of shape (n_obs_sp, n_obs_sc)
+        """
+        return jax.device_get(
+            jax.nn.softmax(self.module.params["mapper_unconstrained"], axis=1)
+        )
+
+    def train(
+        self,
+        max_epochs: int = 1000,
+        use_gpu: Optional[Union[str, int, bool]] = None,
+        lr: float = 0.1,
+        plan_kwargs: Optional[dict] = None,
+    ):
+        """
+        Train the model.
+
+        Parameters
+        ----------
+        max_epochs
+            Number of passes through the dataset.
+        use_gpu
+            Use default GPU if available (if None or True), or index of GPU to use (if int),
+            or name of GPU (if str, e.g., `'cuda:0'`), or use CPU (if False).
+        lr
+            Optimiser learning rate (default optimiser is :class:`~pyro.optim.ClippedAdam`).
+            Specifying optimiser via plan_kwargs overrides this choice of lr.
+        plan_kwargs
+            Keyword args for :class:`~scvi.train.JaxTrainingPlan`. Keyword arguments passed to
+            `train()` will overwrite values present in `plan_kwargs`, when appropriate.
+        """
+        update_dict = {
+            "optim_kwargs": {
+                "learning_rate": lr,
+                "eps": 1e-8,
+                "weight_decay": 0,
+            }
+        }
+        if plan_kwargs is not None:
+            plan_kwargs.update(update_dict)
+        else:
+            plan_kwargs = update_dict
+        device = jax.devices("cpu")[0]
+        if use_gpu is None or use_gpu is True:
+            try:
+                device = jax.devices("gpu")[0]
+                self.module.to(device)
+                logger.info(
+                    "Jax module moved to GPU. "
+                    "Note: Pytorch lightning will show GPU is not being used for the Trainer."
+                )
+            except RuntimeError:
+                logger.debug("No GPU available to Jax.")
+        else:
+            self.module.to(device)
+            logger.info("Jax module moved to CPU.")
+        tensor_dict = self._get_tensor_dict(device=device)
+        training_plan = JaxTrainingPlan(self.module, **plan_kwargs)
+        module_init = self.module.init(self.module.rngs, tensor_dict)
+        state, params = module_init.pop("params")
+        training_plan.set_train_state(params, state)
+        train_step_fn = JaxTrainingPlan.jit_training_step
+        pbar = track(range(max_epochs), style="tqdm", description="Training")
+        history = pd.DataFrame(index=np.arange(max_epochs), columns=["loss"])
```

Comment:
cc @justjhong it's actually pretty flexible

you can use the training plan just as a shell